js/bitopro.js

```diff
@@ -0,0 +1,1151 @@
+            'currency': currency['id'],
+            'amount': this.numberToString (amount),
+            'address': address,
+            'protocol': '',
+        };
+        if (tag !== undefined) {
+            request['message'] = tag;
+        }
+        const response = await this.privatePostWalletWithdrawCurrency (this.extend (request, params));
+        const result = this.safeValue (response, 'data', {});
+        //
+        //     {
+        //         "data":{
+        //             "serial":"20220215BW14069838",
+        //             "currency":"USDT",
+        //             "protocol":"TRX",
+        //             "address":"TKrwMaZaGiAvtXCFT41xHuusNcs4LPWS7w",
+        //             "amount":"8",
+        //             "fee":"2",
+        //             "total":"10"
+        //         }
+        //     }
+        //
+        return this.parseTransaction (result, currency);
+    }
+
+    sign (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
+        let url = '/' + this.implodeParams (path, params);
+        const query = this.omit (params, this.extractParams (path));
+        if (headers === undefined) {
+            headers = {};
+        }
+        headers['X-BITOPRO-API'] = 'ccxt';
+        if (api === 'private') {
+            this.checkRequiredCredentials ();
+            if (method === 'POST') {
+                body = this.json (params);
+                const payload = this.stringToBase64 (body);
+                const signature = this.hmac (payload, this.secret, 'sha384', 'hex');
+                headers['X-BITOPRO-APIKEY'] = this.apiKey;
+                headers['X-BITOPRO-PAYLOAD'] = payload;
+                headers['X-BITOPRO-SIGNATURE'] = signature;
+            } else if (method === 'GET' || method === 'DELETE') {
+                if (Object.keys (query).length) {
+                    url += '?' + this.urlencode (query);
+                }
+                const nonce = this.milliseconds ();
+                let rawData = {
+                    'identity': this.email,
+                    'nonce': nonce,
+                };
+                rawData = this.json (rawData);
+                const payload = this.stringToBase64 (rawData);
+                const signature = this.hmac (payload, this.secret, 'sha384', 'hex');
+                headers['X-BITOPRO-APIKEY'] = this.apiKey;
+                headers['X-BITOPRO-PAYLOAD'] = payload;
+                headers['X-BITOPRO-SIGNATURE'] = signature;
+            }
+        } else if (api === 'public' && method === 'GET') {
+            if (Object.keys (query).length) {
+                url += '?' + this.urlencode (query);
+            }
+        }
+        url = this.urls['api'] + url;
+        return { 'url': url, 'method': method, 'body': body, 'headers': headers };
+    }
+
+    handleErrors (code, reason, url, method, headers, body, response, requestHeaders, requestBody) {
+        if (code >= 200 && code < 300) {
```

Comment:
![image](https://user-images.githubusercontent.com/43336371/154237552-bca41717-780b-40e9-b87d-1550d30d6ed5.png)

You should try to understand whenever we got an errored response from the exchange, parse that error and match (broadly or exactly) it with the adequate exception (following our error hierarchy). In this case 

n this case, you could extract "Invalid amount 0.1 less than min 1000.000000." and broadly match it with the InvalidOrder exception

Rough example:

First, add the exceptions object to the exchange definition (top of the file)

            'exceptions': {
                'exact': {...}
                'broad': {
                    'Invalid amount less than min': InvalidOrder, //  {"error":"Invalid amount 0.1 less than min 1000.000000."}


Then here in the handleErrors:

```
const error = this.safeValue(response, 'error);
if (error !== undefined) {
               const feedback = this.id + ' ' + body;
                this.throwExactlyMatchedException (this.exceptions['exact'], error, feedback);
                this.throwBroadlyMatchedException (this.exceptions['broad'], body, feedback);
}
```

Check other exchanges for reference 🙂